### DISCUSSION TITLE
Polyfill illegal opcodes with a VxD

### DISCUSSION THREAD

**Discussion by GravisZro** [itsmattkc/dotnet9x#29] [Ideas]
### Problem:
While the rogue SSE2 instruction may have been a one-off in the installer, you are going to run into a LOT of instructions that aren't implemented when running .Net 2.0 applications and especially later versions. This means the current method of patching executables is **unsustainable** and will be increasingly hacky when there are integrity checks for the binary.

### Solution:
Instead of getting rid of unknown instructions, you should [polyfill](https://en.wikipedia.org/wiki/Polyfill_(programming)) them. This is possible because on Windows 95 executing an illegal operation causes an [Invalid Opcode Exception](https://wiki.osdev.org/Exceptions) which is handle as such: [IDT](https://wiki.osdev.org/Interrupt_Descriptor_Table) -> VMM -> VPICD (Virtual Programmable Interrupt Controller Device). Windows 95's VPICD service allows you to install an interrupt specific handler using the `VPICD_Virtualize_IRQ` function.

To make a polyfilling VxD then someone would need the [Windows 95 DDK](https://archive.org/details/Microsoft_Developer_Network_DDK_Windows_95_DDK_Disc_9_October_1995) and a PDF with info like [Writing Windows VxDs and Device Drivers (1996)](http://libgen.li/ads.php?md5=856746818e92a68a6b4b0a7b7a2348a5) (click "GET").

**Comment by itsmattkc** [comment]
> While the rogue SSE2 instruction may have been a one-off in the installer, you are going to run into a LOT of instructions that aren't implemented when running .Net 2.0 applications and especially later versions. This means the current method of patching executables is **unsustainable** and will be increasingly hacky when there are integrity checks for the binary.

I think there's a misunderstanding here, and looking back it probably wasn't clear in the video. I didn't patch out an SSE2 _instruction_, I patched an erroneous check .NET's JIT compiler did that was causing it to emit SSE2 code on a platform that didn't support it. I don't suspect SSE2 will be an issue for .NET (at least 2.0-3.5) moving forward.

That being said, your idea could be interesting if this gets spun off into a more general kernel extender for apps outside of .NET, and would be worth looking into at that point.

**Reply by GravisZro** [reply]
A misunderstanding about the a the finer details of a highly technical point happened on the internet? Seems super unlikely but there is a first time for everything.

Anyway, this is actually an **old** trick (1970s) that can be used for all sorts of abuses. I think the polyfill VxD is vital because my understanding is that few .NET applications are actually MSIL.

There is also the matter of the system requirements for Win95 only being an i386 which several generations away from even supporting SSE. I'm not saying it's dire feature because I'm a time traveler trying to impress my friends in 1985 but I just feel like if you're going do something then do it right. Also, if you so happen to be a time traveler then you totally could impress your friends in 1985, not that I would break the Timeline Accords for something so trivial because I wouldn't.
